cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java

```diff
@@ -205,15 +208,53 @@ public void onClose(KubernetesClientException e) {
         }
     }
 
+    private static ObjectMeta templateMetadata(StatefulSet resource) {
+        return resource.getSpec().getTemplate().getMetadata();
+    }
+
     public String getPodName(StatefulSet desired, int podId) {
-        return desired.getMetadata().getName() + "-" + podId;
+        return templateMetadata(desired).getName() + "-" + podId;
+    }
+
+    private void setGeneration(StatefulSet desired, int nextGeneration) {
+        templateMetadata(desired).getAnnotations().put(ANNOTATION_GENERATION, String.valueOf(nextGeneration));
+    }
+
+    private static int getGeneration(ObjectMeta objectMeta) {
+        if (objectMeta.getAnnotations().get(ANNOTATION_GENERATION) == null) {
+            new RuntimeException().printStackTrace();
+        }
+        return Integer.parseInt(objectMeta.getAnnotations().getOrDefault(ANNOTATION_GENERATION, "-2"));
+    }
+
+    protected void incrementGeneration(StatefulSet current, StatefulSet desired) {
+        final int generation = Integer.parseInt(templateMetadata(current).getAnnotations().getOrDefault(ANNOTATION_GENERATION, "1"));
+        final int nextGeneration = generation + 1;
+        setGeneration(desired, nextGeneration);
+    }
+
+    protected abstract boolean shouldIncrementGeneration(StatefulSet current, StatefulSet desired);
+
+    private static int getSsGeneration(StatefulSet resource) {
+        if (resource == null) {
+            return -3;
+        }
+        return getGeneration(templateMetadata(resource));
+    }
+
+    private static int getPodGeneration(Pod resource) {
+        if (resource == null) {
+            return -4;
+        }
+        return getGeneration(resource.getMetadata());
     }
 
     @Override
-    protected Future<ReconcileResult<P>> internalCreate(String namespace, String name, StatefulSet desired) {
+    protected Future<ReconcileResult<StatefulSet>> internalCreate(String namespace, String name, StatefulSet desired) {
         // Create the SS...
-        Future<ReconcileResult<P>> result = Future.future();
-        Future<ReconcileResult<P>> crt = super.internalCreate(namespace, name, desired);
+        Future<ReconcileResult<StatefulSet>> result = Future.future();
+        setGeneration(desired, 1);
```

Comment:
Just a thought ... if we start from generation 0 on creation, the next generation values should represent the number of rolling updates so far. Maybe it's useful to know ... maybe not.

Reply:
On the basis that annotations are more private impl detail than public API I think it's probably right that for those annotation which are impl detail purely of the CO, that they're scoped to the CO.